docs/develop/create-sso-office-add-ins-nodejs.md

```diff
@@ -1,14 +1,574 @@
+    ```
+
+1. On rare occasions the bootstrap token that Office has cached is unexpired when Office validates it, but expires by the time it reaches Azure AD for exchange. Azure AD will respond with error **AADSTS500133**. In this case, the add-in should simply recursively recall `getGraphData`. Since the cached bootstrap token is now expired, Office will get a new one from Azure AD. The So replace `TODO 8` with the following. 
+
+    ```javascript
+    if (exchangeResponse.error_description.indexOf("AADSTS500133") !== -1)       
+    {
+        getGraphData();
+    }
+    ```
+
+1. To ensure that the add-in doesn't enter an infinite loop of calls to `getGraphData`, the add-in should keep track of how many times `getGraphData` has been called and be sure that is not called recursively called more than once. So, create a counter variable in a scope that is global to the `handleAADErrors` and `getGraphData` functions. A good place for global variables is just below the call of the `Office.onReady` method. 
+
+    ```javascript
+    let retryGetAccessToken = 0;
+    ```
+
+1. Change the `if` structure in the `handleAADErrors` method so that it:
+
+    - Increments the counter just before it calls `getGraphData`.
+    - Tests to ensure that `getGraphData` has not already been called a second time. 
+
+    So the final version of the `if` structure should look like the following:
+
+    ```javascript
+    if ((exchangeResponse.error_description.indexOf("AADSTS500133") !== -1)
+        &&
+        (retryGetAccessToken <= 0)) 
+    {
+        retryGetAccessToken++;
+        getGraphData();
+    }
+    ```
+
+1. Replace `TODO 9` with the following. 
+
+    ```javascript
+    else 
+    {                
+        dialogFallback();
+    }
+    ```
+
+1. Save and close the file.
+
+### Get the data and add it to the Office document
+
+1. In the `public\javascripts` folder, create a new file named `data.js`.
+
+1. Add the following function to the file. This is the function that is called by the `getGraphData` function when it has acquired an access token to Microsoft Graph. 
+
+    ```javascript
+    function makeGraphApiCall(accessToken) {
+        $.ajax(
+
+            // TODO 10: Call an Express route on the add-in's server-side 
+            //          code and pass the access token to Microsoft Graph.
+
+        )
+        .done(function (response) {
+
+            // TODO 11: Write the data received from Microsoft Graph to 
+            //          the Office document.
+
+        })
+        .fail(function (errorResult) {
+            showMessage("Error from Microsoft Graph: " + JSON.stringify(errorResult));
+        });
+    }
+    ```
+
+1. Replace `TODO 10` woth the following. About this code, note: 
+
+    - This object is the parameter to the `$.ajax` method.
+    - The `/getuserdata` is an Express route on the add-in's server that you create in a later step. It will call a Microsoft Graph endpoint and include the access token in it's call. 
```

Comment:
```suggestion
    - The `/getuserdata` is an Express route on the add-in's server that you create in a later step. It will call a Microsoft Graph endpoint and include the access token in its call. 
```